src/common/TrackedOp.cc

```diff
@@ -265,28 +265,29 @@ bool OpTracker::register_inflight_op(TrackedOp *i)
   return true;
 }
 
-void OpTracker::unregister_inflight_op(TrackedOp *i)
+void OpTracker::unregister_inflight_op(TrackedOpRef&& i)
 {
+  const auto ri = i.get();
   // caller checks;
-  assert(i->state);
+  assert(ri->state);
 
-  uint32_t shard_index = i->seq % num_optracker_shards;
+  uint32_t shard_index = ri->seq % num_optracker_shards;
   ShardedTrackingData* sdata = sharded_in_flight_list[shard_index];
   assert(NULL != sdata);
   {
     Mutex::Locker locker(sdata->ops_in_flight_lock_sharded);
-    auto p = sdata->ops_in_flight_sharded.iterator_to(*i);
+    auto p = sdata->ops_in_flight_sharded.iterator_to(*ri);
     sdata->ops_in_flight_sharded.erase(p);
   }
-  i->_unregistered();
+  ri->_unregistered();
 
   if (!tracking_enabled)
-    delete i;
+    delete i.detach();
   else {
     RWLock::RLocker l(lock);
-    i->state = TrackedOp::STATE_HISTORY;
+    ri->state = TrackedOp::STATE_HISTORY;
     utime_t now = ceph_clock_now();
-    history.insert(now, TrackedOpRef(i));
+    history.insert(now, std::move(i));
   }
 }
```

Comment:
Instead of this mess of changes to unregister_inflight_op(), can we just move everything from if (!tracking_enabled) into the caller?  There is only one caller, and having this *inside* the function makes its name misleading (it's not just unregistering but conditionally either deleting or reregistering elsewhere).  I find it very unsettling to have a ref to the object (ri at the top) and also detach+delete.

Reply:
Yeah, having `delete` and the re-registration part is uncomfortable at best even without the extra complexity of `Ref`. I'm going to make a small clean-up in a separated commit for the sake of backportability.